core/policy/repository.go

```diff
@@ -0,0 +1,212 @@
+
+		var err error
+		if err = pr.fetchPolicyRules(&policyListNew[index]); err != nil {
+			return errors.Wrap(err, "initial fetch failed")
+		}
+	}
+
+	pr.policyList = policyListNew
+	return nil
+}
+
+// RulesForPolicy gives rules of given policy
+func (pr *Repository) RulesForPolicy(policy market.AccessPolicy) (market.AccessPolicyRuleSet, error) {
+	pr.policyLock.RLock()
+	defer pr.policyLock.RUnlock()
+
+	index, exist := pr.getPolicyIndex(pr.policyList, policy)
+	if !exist {
+		return market.AccessPolicyRuleSet{}, fmt.Errorf("unknown policy: %s", policy)
+	}
+
+	return pr.policyList[index].rules, nil
+}
+
+// RulesForPolicies gives list of rules of given policies
+func (pr *Repository) RulesForPolicies(policies []market.AccessPolicy) ([]market.AccessPolicyRuleSet, error) {
+	pr.policyLock.RLock()
+	defer pr.policyLock.RUnlock()
+
+	policiesRules := make([]market.AccessPolicyRuleSet, len(policies))
+	for i, policy := range policies {
+		index, exist := pr.getPolicyIndex(pr.policyList, policy)
+		if !exist {
+			return []market.AccessPolicyRuleSet{}, fmt.Errorf("unknown policy: %s", policy)
+		}
+		policiesRules[i] = pr.policyList[index].rules
+	}
+	return policiesRules, nil
+}
+
+func (pr *Repository) getPolicyIndex(policyList []policyMetadata, policy market.AccessPolicy) (int, bool) {
+	for index, policyMeta := range policyList {
+		if policyMeta.policy == policy {
+			return index, true
+		}
+	}
+
+	return 0, false
+}
+
+func (pr *Repository) fetchPolicyRules(policyMeta *policyMetadata) error {
+	req, err := requests.NewGetRequest(policyMeta.policy.Source, "", nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to create policy request")
+	}
+	req.Header.Add("If-None-Match", policyMeta.eTag)
+
+	res, err := pr.client.Do(req)
+	if err != nil {
+		return errors.Wrapf(err, "failed fetch policy rule %s", policyMeta.policy)
+	}
+	defer res.Body.Close()
+
+	httptrace.TraceRequestResponse(req, res)
+
+	if res.StatusCode == http.StatusNotModified {
+		return nil
+	}
+	if err := requests.ParseResponseError(res); err != nil {
+		return errors.Wrapf(err, "failed to fetch policy rule %s", policyMeta.policy)
+	}
+
+	policyMeta.rules = market.AccessPolicyRuleSet{}
+	err = requests.ParseResponseJSON(res, &policyMeta.rules)
+	if err != nil {
+		return errors.Wrapf(err, "failed to parse policy rule %s", policyMeta.policy)
+	}
+	policyMeta.eTag = res.Header.Get("ETag")
+	return nil
+}
+
+func (pr *Repository) fetchLoop() {
+	for {
+		select {
+		case <-pr.fetchShutdown:
+			return
+		case <-time.After(pr.fetchInterval):
+			pr.policyLock.Lock()
+			policyListActive := make([]policyMetadata, len(pr.policyList))
+			copy(policyListActive, pr.policyList)
+			pr.policyLock.Unlock()
+
+			for index := range policyListActive {
+				var err error
+				if err = pr.fetchPolicyRules(&policyListActive[index]); err != nil {
+					log.Warn().Err(err).Msg("synchronise fetch failed")
+				}
+
+				pr.policyLock.Lock()
+				pr.policyList[index] = policyListActive[index]
```

Comment:
This could create some nasty bugs. You updating here original policyList item by index of copy of policy, but in theory at this time original pr.policyList may change and it's indexes will not mach any more so you will end up updating wrong item.

I suggest modify `AddPolicies` so it only adds request to channel and then you can handle it here if fetchLoop so you have only once place which modifies pr.policyList.

```go
func (pr *Repository) fetchLoop() {
	for {
		select {
		case <-pr.fetchShutdown:
                        return;
                case policies := <-pr.newPolicies:
                       // Handle new policies
                case <-time.After(pr.fetchInterval):
                      // ...
 
```